myshop/tests/shop/test_views.py

```diff
@@ -0,0 +1,62 @@
+from django.test import TestCase
+from django.urls import reverse
+from tests.shop.test_model_mixin_testcase import ModelMixinTestCase
+from shop.models import Product
+
+
+class TestViews(ModelMixinTestCase, TestCase):
+    def test_templates_used_for_product_list_view(self):
+        product_list_url = reverse("shop:product_list")
+        response = self.client.get(product_list_url)
+
+        self.assertTemplateUsed(response, "shop/product/list.html")
+
+    def test_templates_used_for_product_list_view_with_category_slug(self):
+        product_list_by_category_url = reverse(
+            "shop:product_list_by_category", args=[self.category1.slug]
+        )
+        response = self.client.get(product_list_by_category_url)
+
+        self.assertTemplateUsed(response, "shop/product/list.html")
+
+    def test_templates_used_for_product_detail_view(self):
+        product_list_url = reverse(
+            "shop:product_detail",
+            args=[self.product1.id, self.product1.slug],
+        )
+        response = self.client.get(product_list_url)
+
+        self.assertTemplateUsed(response, "shop/product/detail.html")
+
+    def test_product_detail_returns_404_for_invalid_slug(self):
+        product_detail_url = reverse(
+            "shop:product_detail",
+            args=[self.product1.id, "invalid_product_slug"],
+        )
+        response = self.client.get(product_detail_url)
+
+        self.assertEquals(404, response.status_code)
+
+    def test_post_list_shows_all_product_when_category_not_selected(self):
+        all_products = Product.objects.all()
+        request = self.client.get(reverse("shop:product_list"))
+        response = request.context.get("products")

```

Comment:
Incorrect naming here. It should be products. Any how you can use the `request.context.get("products")` directly and avoid inline temp

Reply:
Resolved by renaming and using context direct rather than inline variable